ultimarc/devices/_device.py

```diff
@@ -222,22 +218,34 @@ def read(self, b_request, w_value, w_index, data=None, size=None, request_type=u
 
         # Combine direction, request type and recipient together.
         request_type = usb.LIBUSB_ENDPOINT_IN | request_type | recipient
-        # we need to add 8 extra bytes to the data buffer for read requests.
+        # we need to add 8 extra bytes to the structure buffer for read requests.
         return self._make_control_transfer(request_type, b_request, w_value, w_index, data, size)
 
-    def read_interrupt(self, endpoint, data, length, actual_length):
+    def read_interrupt(self, endpoint, response, uses_report_id=True):
         """
-        Read data from USB device on the interrupt endpoint.
-        :param endpoint: Endpoint to receive data on.
-        :param data: Structure to hold data.
-        :param length: Expected length of data to receive.
-        :param actual_length: Actual length received.
+        Read response from USB device on the interrupt endpoint.
+        :param endpoint: Endpoint to receive message on.
+        :param response: Variable holding the complete response from the device.
+        :param uses_report_id: True if report_id is in the message.
         """
         if not self.interface:
             raise USBDeviceInterfaceNotClaimedError(self.dev_key)
 
-        payload_ptr = ct.byref(data)
-        return self._make_interrupt_transfer(endpoint, payload_ptr, length, actual_length)
+        actual_length = int(0)
+        length = 5 if uses_report_id else 4
+        payload = (ct.c_ubyte * length)(0)
+        payload_ptr = ct.byref(payload)
+
+        for pos in range(0, ct.sizeof(response), 4):
+            self._make_interrupt_transfer(endpoint, payload_ptr, length, actual_length)
+            # Remove report_id (byte 0) if it is used
+            ct.memmove(ct.addressof(response)+pos,
+                       ct.byref(payload, 1) if uses_report_id else ct.byref(payload),
+                       4)
```

Comment:
It seems to me that the var 'actual_length' should be used here instead of the hard coded value of 4.  Should there be a check to see if 'actual_length' is less than expected?

Reply:
Yes you are correct.  Updated to use actual_length and add comments about why and when 4 and 5 length sizes are used.